codec-http/src/main/java/io/netty/handler/codec/http/multipart/DefaultHttpDataFactory.java

```diff
@@ -209,67 +225,99 @@ public Attribute createAttribute(HttpRequest request, String name, String value)
             checkHttpDataSize(fileUpload);
-            List<HttpData> fileToDelete = getList(request);
-            fileToDelete.add(fileUpload);
+            List<HttpData> list = getList(request);
+            list.add(fileUpload);
             return fileUpload;
         }
+
         if (checkSize) {
-            FileUpload fileUpload = new MixedFileUpload(name, filename, contentType,
-                    contentTransferEncoding, charset, size, minSize);
+            FileUpload fileUpload = new MixedFileUpload(
+                    name, filename, contentType, contentTransferEncoding, charset, size, minSize);
             fileUpload.setMaxSize(maxSize);
             checkHttpDataSize(fileUpload);
-            List<HttpData> fileToDelete = getList(request);
-            fileToDelete.add(fileUpload);
+            List<HttpData> list = getList(request);
+            list.add(fileUpload);
             return fileUpload;
         }
-        MemoryFileUpload fileUpload = new MemoryFileUpload(name, filename, contentType,
-                contentTransferEncoding, charset, size);
+
+        MemoryFileUpload fileUpload = new MemoryFileUpload(
+                name, filename, contentType, contentTransferEncoding, charset, size);
         fileUpload.setMaxSize(maxSize);
         checkHttpDataSize(fileUpload);
         return fileUpload;
     }
 
     @Override
     public void removeHttpDataFromClean(HttpRequest request, InterfaceHttpData data) {
-        if (data instanceof HttpData) {
-            List<HttpData> fileToDelete = getList(request);
-            fileToDelete.remove(data);
+        if (!(data instanceof HttpData)) {
+            return;
+        }
+
+        // Do not use getList because it adds empty list to requestFileDeleteMap
+        // if request is not found
+        List<HttpData> list = requestFileDeleteMap.get(request);
+        if (list == null) {
+            return;
+        }
+
+        // Can't simply call list.remove(data), because different data items may be equal.
+        // Need to check identity.
+        Iterator<HttpData> i = list.iterator();
+        while (i.hasNext()) {
+            HttpData n = i.next();
+            if (n == data) {
+                i.remove();
+                break;
+            }
+        }
+
+        // Remove empty list to avoid memory leak
+        if (list.isEmpty()) {
+            requestFileDeleteMap.remove(request);
         }
     }
 
     @Override
     public void cleanRequestHttpData(HttpRequest request) {
-        List<HttpData> fileToDelete = requestFileDeleteMap.remove(request);
-        if (fileToDelete != null) {
-            for (HttpData data: fileToDelete) {
-                data.delete();
-            }
-            fileToDelete.clear();
+        List<HttpData> list = requestFileDeleteMap.remove(request);
+        if (list == null) {
+            return;
+        }
+
+        for (HttpData data: list) {
+            data.release();
         }
+        list.clear();
```

Comment:
is this necessary? can we just let it be GCed instead of clearing it out first?

Reply:
Removed `list.clear()`